File: 47.permutations-ii.cpp
/*
 * @lc app=leetcode id=47 lang=cpp
 *
 * [47] Permutations II
 */
#include <iostream>
#include <vector>
using namespace std;
// @lc code=start
// Time: O(n!)
// Space: O(n)
class Solution {
public:
    vector<vector<int> > permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> path;
        vector<vector<int> > res;
        dfs(nums, path, res);
        return res;
    }
    void dfs(vector<int>& nums, vector<int> path, vector<vector<int> >& res){
        if (nums.empty() || nums.size() == 0){
            res.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); i++){
            if(i > 0 && nums[i-1] == nums[i]){
                continue;
            }
            path.push_back(nums[i]);
            vector<int> tmp = nums;
            tmp.erase(tmp.begin()+i);
            dfs(tmp, path, res);
            path.pop_back();
        }
    }
};
// 30/30 cases passed (28 ms)
// Your runtime beats 43.33 % of cpp submissions
// Your memory usage beats 14.28 % of cpp submissions (11.9 MB)
// @lc code=end

void print(vector<vector<int> > const &input)
{
    for (int i = 0; i < input.size(); i++) {
        for (int j = 0; j < input[i].size(); j++){
            cout << input[i][j] << ' ';
        }
        cout << endl;
    }
}
int main(){
    Solution* sol = new Solution();
    vector<int> test = {1, 1, 2};
    vector<vector<int> > res = sol->permuteUnique(test);
    print(res);
}

File: LeetCode/47.permutations-ii.py
#
# @lc app=leetcode id=47 lang=python3
#
# [47] Permutations II
#
from typing import List
# @lc code=start
# Time: O(n^2)
# Space: O(n)
# class Solution:
#     def permuteUnique(self, nums: List[int]) -> List[List[int]]:
#         if not nums:
#             return []
#         res = []
#         nums.sort()
#         self.dfs(nums, [], res)
#         return res
    
#     def dfs(self, nums, path, res):
#         if not nums or len(nums) == 0:
#             res.append(path)
#         n = len(nums)
#         for i in range(n):
#             if i >= 1 and nums[i-1] == nums[i]:
#                 continue
#             self.dfs(nums[:i]+nums[i+1:], path+[nums[i]], res)
# # 30/30 cases passed (48 ms)
# # Your runtime beats 98.52 % of python3 submissions
# # Your memory usage beats 100 % of python3 submissions (12.8 MB)

# 2. iterative
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        for n in nums:
            tmp = []
            for item in res:
                for i in range(len(item)+1):
                    tmp.append(item[:i]+[n]+item[i:])
                    if i < len(item) and item[i] == n:
                        break
            res = tmp
        return res
# 30/30 cases passed (48 ms)
# Your runtime beats 98.52 % of python3 submissions
# Your memory usage beats 100 % of python3 submissions (12.9 MB)
# @lc code=end

print(Solution().permuteUnique([1,1,2]))
print(Solution().permuteUnique([3,3,0,3]))